syntax = "proto3";

message File
{
    string file_name = 1;
    repeated Struct struct_list = 2;
    repeated Function function_list = 3;
}

message Struct 
{
    repeated string namespace = 1;
    string name = 2;

    repeated Method methods = 3;
    repeated Variable variables = 4; 
    Function constructor = 5;
}

message Method
{
    Access access = 1;
    Function function = 2;
}

message Function
{
    repeated string namespace = 1;
    string name = 2;
    string output_param = 3;
    repeated Param input_params = 4;
}

message Variable
{
    Access access = 1;
    Param variable = 2;
}

message Param 
{
    string name = 1;
    string type = 2;
}

enum Access {
    NONE = 0;
    PUBLIC = 1;
    PRIVATE = 2;
    PROTECTED = 3;
}